src/cms/logic.py

```diff
@@ -1,15 +1,367 @@
+                continue
+            part_id = part.get('id', '')
+            if part_id:
+                part_url = urljoin(url, '#' + part_id)
+            else:
+                part_url = remove_fragment(url)
+            part_name = part.get_text().strip()
+            part_text = get_text_for_header(part)
+            docs = add_part_as_doc(docs, part_url, part_name, part_text)
+            part.decompose()
+    return docs
+
+
+def get_page(
+    url: str,
+    fetched_urls: set[str]
+) -> Tuple[BeautifulSoup | None, set[str]]:
+    time.sleep(.1)
+    try:
+        fetched_urls.add(url)
+        headers = {
+            'Accept': 'text/html; charset=utf-8'
+        }
+        response = requests.get(url, headers=headers)
+    except requests.exceptions.ConnectionError:
+        logger.warn(f'Could not access {url}')
+        logger.warn('Please run server to index site search')
+        return None, fetched_urls
+
+    if response.status_code != 200:
+        logger.warn(f'Could not access {url}')
+        return None, fetched_urls
+
+    if 'text/html' not in response.headers['Content-Type']:
+        return None, fetched_urls
+
+    soup = BeautifulSoup(response.text, 'html.parser')
+    return soup, fetched_urls
+
+
+def get_name(html: BeautifulSoup) -> str:
+    try:
+        return html.find('h1').get_text().strip()
+    except AttributeError:
+        return html.title.get_text().strip()
+
+
+def get_body(html: BeautifulSoup) -> Tag | None:
+    body = html.find('body')
+    if not isinstance(body, Tag):
+        return
+    for non_content_selector in ['script', '.djdt-hidden']:
+        for element in body.select(non_content_selector):
+            element.decompose()
+    for element in body.find_all(text=lambda text: isinstance(text, Comment)):
+        element.extract()
+    return body
+
+
+@cache(900)
+def get_base(press: press_models.Press) -> str:
+    return press.site_url()
+
+
+@cache(900)
+def excluded_urls() -> List[str]:
+    return [
+        journal.site_url() for journal in journal_models.Journal.objects.all()
+    ] + [
+        repo.site_url() for repo in repository_models.Repository.objects.all()
+    ]
+
+
+def url_in_scope(press: press_models.Press, deeper_url: str) -> bool:
+    base = get_base(press)
+    if urlparse(deeper_url).hostname != urlparse(base).hostname:
+        return False
+    for excluded_url in excluded_urls():
+        if excluded_url in deeper_url:
+            return False
+    return True
+
+
+def remove_fragment(url: str) -> str:
+    fragment = urlparse(url).fragment
+    return url.replace('#' + fragment, '')
+
+
+def normalize_url(url: str) -> str:
+    url = remove_fragment(url)
+    if url.endswith('/'):
+        url = url[:-1]
+    return url
+
+
+def url_is_new(fetched_urls: set[str], deeper_url: str) -> bool:
+    if deeper_url in fetched_urls:
+        return False
+    for fetched_url in fetched_urls:
+        if normalize_url(fetched_url) == normalize_url(deeper_url):
```

Comment:
minor improvement:
if we stored the result of `normalize_url` in `fetched_urls` we would avoid the *O(N \* M)* time complexity of having to iterate the whole set everytime `url_is_new` is called